src/cpp/shared/lib/optional.hpp

```diff
@@ -0,0 +1,1222 @@
+//
+// Copyright (c) 2014-2018 Martin Moene
+//
+// https://github.com/martinmoene/optional-lite
```

Comment:
SMH that something 1200 lines long is the "lite" version.  Thanks C++...

Also, just curious, which version of the standard are we targeting here?  `std::optional` is available in C++17, is that possible to use here?

Reply:
We're running on Visual Studio 2015, which has only some of the C++11 features. I don't remember all the barriers to upgrading, but we decided a while ago that it was a no-go.